### PR TITLE
Optimise getFileInfo - for mounts, dont fetch again filecache, but reuse sourceRootInfo

### PR DESCRIPTION
**Pull request by mrow4a**
- [ ] Optimization
- [ ] Unit tests

### CODE REVIEW

**Comment by mrow4a** [comment]
Ok, I see it wont be that easy looking on unit tests

File: apps/files_sharing/lib/IShareCache.php [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+namespace OCA\Files_Sharing;
+
+
+interface IShareCache {
```
Comment (jvillafanez) [inline apps/files_sharing/lib/IShareCache.php]: I'm not sure about adding this new interface.

First, the name seems wrong. There is already a cache interface available, so when I implement this `IShareCache` I expect to implement the methods for a cache (get, set, remove, etc). Right now we only have a `getMetadata` method which doesn't fit being a cache by itself.

Second, the `getMetadata` method usage seems bound to the cache. Does it make sense to implement an isolated `getMetadata` method? I don't think so.

I think the cleanest option is to create a new interface (`ICacheMetadata` or something similar) in the OCP namespace that inherit from the public `ICache` interface (I assume there is such interface) and includes the new `getMetadata` method. The problem with this solution is that we'll likely need to rewrite these caches.

Another option could be forget this interface and just add the new method.

File: lib/private/Files/View.php [hunk 2]
```diff
@@ -1365,7 +1365,11 @@ public function getFileInfo($path, $includeMountPoints = true) {
 								continue;
 							}
 							$subCache = $subStorage->getCache('');
-							$rootEntry = $subCache->get('');
+							if ($parentReadOnly && $subCache instanceof \OCA\Files_Sharing\IShareCache) {
```
Comment (jvillafanez) [inline lib/private/Files/View.php]: 🚫 core shouldn't have any knowledge about specific apps.

File: lib/private/Files/View.php [hunk 3]
```diff
@@ -1430,9 +1434,18 @@ public function getDirectoryContent($directory, $mimetype_filter = '') {
 				$mountPoint = $mount->getMountPoint();
 				$subStorage = $mount->getStorage();
 				if ($subStorage) {
+
 					$subCache = $subStorage->getCache('');
+					if ($subCache instanceof \OCA\Files_Sharing\IShareCache) {
```
Comment (jvillafanez) [inline lib/private/Files/View.php]: Same as above

**Comment by mrow4a** [comment]
@jvillafanez This PR is purely experimental, I do not try to revolutionize anything (since the target is for OC10, not OC11), but optimize using existing structure. Problem there is that https://github.com/owncloud/core/blob/master/lib/private/Files/View.php#L1431 `getStorage()` will fetch data from oc_filecache, then https://github.com/owncloud/core/blob/master/lib/private/Files/View.php#L1435 we will exactly do the same in case of SharedMounts. 

The other problem is that https://github.com/owncloud/core/blob/master/lib/private/Files/View.php#L1433 is generaly a free ride, this just extends Cache. The check you mentioned `if ($parentReadOnly && $subCache instanceof \OCA\Files_Sharing\IShareCache) {` is just for "debugging" purposes, just to try to avoid unnecesary calls in some cases. The problem is this code is "general purpose" and handles any scenario. 

The flag for this PR is Developing, but any feedback is generally most needed, thanks!

File: apps/dav/lib/Connector/Sabre/Directory.php [hunk 4]
```diff
@@ -261,7 +261,7 @@ public function getChildren() {
 				// the caller believe that the collection itself does not exist
 				throw new Forbidden('No read permissions');
 			}
-			$folderContent = $this->fileView->getDirectoryContent($this->path);
+			$folderContent = $this->fileView->getDirectoryContent($this->path, '', true);
```
Comment (mrow4a) [inline apps/dav/lib/Connector/Sabre/Directory.php]: This is default value for this function

File: apps/files_sharing/lib/Cache.php [hunk 5]
```diff
@@ -73,6 +73,15 @@ public function __construct($storage, IStorage $sourceStorage, ICacheEntry $sour
 		);
 	}
 
+	/**
+	 * Obtain metadata for this share mount
+	 * 
+	 * @return array
+	 */
+	public function getMetadata() {
+		return $this->formatCacheEntry($this->sourceRootInfo);
```
Comment (PVince81) [inline apps/files_sharing/lib/Cache.php]: not sure what you're trying to achieve but something that comes to mind is `getSourceCache()->get('')`

**Pull request closed without merging**